poc/observe.py

```diff
@@ -0,0 +1,689 @@
+from functools import partial
+import itertools
+import logging
+import threading
+
+from traits import ctraits
+from traits.constants import ComparisonMode
+from traits.trait_base import Undefined, Uninitialized
+from traits.ctrait import CTrait
+from traits.trait_dict_object import TraitDictObject
+from traits.trait_set_object import TraitSetObject
+
+
+from poc.events import CTraitObserverEvent, ListObserverEvent
+from poc.interfaces import INotifiableObject
+from poc.trait_observer_notifier import (
+    TraitObserverNotifier,
+)
+from poc.listener_change_notifier import (
+    ListenerChangeNotifier,
+)
+
+# We need to identify objects which has this `_notifiers` methods
+# We could do the easy-to-ask-forgiveness-than-permission way.
+# Or we could do the leap-before-you-leap way.
+# Here is the LBYL way
+from traits.trait_list_object import TraitListObject
+INotifiableObject.register(TraitListObject)
+INotifiableObject.register(CTrait)
+INotifiableObject.register(ctraits.CHasTraits)
+
+logger = logging.getLogger()
+
+
+def dispatch_same(callback, args=(), kwargs=None):
+    if kwargs is None:
+        kwargs = {}
+    callback(*args, **kwargs)
+
+
+def dispatch_new_thread(callback, args=(), kwargs=None):
+    if kwargs is None:
+        kwargs = {}
+    threading.Thread(
+        target=callback, args=args, kwargs=kwargs).start()
+
+
+# These are user-facing notifier factories...
+DISPATCHERS = {
+    "extended": dispatch_same,
+    "same": dispatch_same,
+    "new": dispatch_new_thread,
+}
+
+
+def observe(object, callback, path, remove, dispatch):
+    """
+    Parameters
+    ----------
+    object : HasTrait
+        An object that implements `_notifiers` for returning a list for
+        adding or removing notifiers
+    callback : callable(object)
+        A callable which will receive an event object with information
+        about a change. The type and interface of this event object
+        depends on the type of events. For example, a change to a simple
+        trait will emit CTraitObserverEvent. A mutation to a list
+        will emit ListObserverEvent.
+    path : ListenerPath
+        Path for listening to extended traits.
+    remove : boolean
+        Whether to remove the observer.
+    dispatch : str
+        A string indicating the thread on which notifications should be
+        run.
+    """
+    if remove:
+        remove_notifiers(
+            path=path,
+            object=object,
+            callback=callback,
+            target=object,
+            dispatcher=DISPATCHERS[dispatch],
+        )
+    else:
+        add_notifiers(
+            path=path,
+            object=object,
+            callback=callback,
+            target=object,
+            dispatcher=DISPATCHERS[dispatch],
+        )
+
+
+def add_notifiers(object, callback, path, target, dispatcher):
```

Comment:
Could you add `target` to the docstring? What's the difference in roles of `object` and `target`?

Reply:
Yes. I had a (hopefully-better) description in `ListenerChangeNotifier`, which I can copy over here:
```
        target : Any
            Sets the context for the (actual) callback. In practice, it is
            the HasTrait instance on which the listener path is defined and
            will be seen by users as the "owner" of the notifier.
            Strictly speaking, this object sets the context for the notifier
            and does not have to be a notifiable object.
```

Reply:
> What's the difference in roles of object and target?

`object` is where the notifiers are going to be attached to. `target` is the context.
In `observer`, `object` and `target` are the same, but as we walk down the path to nested objects, `object` is the nested object onto which notifiers are attached, whereas `target` is kept unchanged (the mother object if this is a better name?).

Reply:
> In observer,

Sorry, in `poc.observe.observe`, not `observer`.